modules/dnn/src/dnn.cpp

```diff
@@ -1357,7 +1358,8 @@ struct Net::Impl
 
         std::vector<LayerPin> pinsForInternalBlobs;
         blobManager.allocateBlobsForLayer(ld, layerShapesIt->second, pinsForInternalBlobs,
-                                          preferableBackend == DNN_BACKEND_INFERENCE_ENGINE);
+                                          preferableBackend == DNN_BACKEND_INFERENCE_ENGINE,
+                                          preferableTarget == DNN_TARGET_OPENCL_FP16);
```

Comment:
We need to allocate halfs if `preferableTarget == DNN_TARGET_OPENCL_FP16` and `preferableBackend == DNN_BACKEND_DEFAULT` because there is one more backend (Intel's Inference Engine) which supports FP16 computations but accepts inputs and outputs in FP32.